R/makeHeatmap.R

```diff
@@ -59,48 +74,95 @@
+
+    if (!is.null(metricInfo)) {
+        metricInfo <- metricInfo[match(colnames(mat), 
+                                       metricInfo[[metricCol]]), , 
+                                 drop = FALSE] %>% as.data.frame()
+        rownames(metricInfo) <- metricInfo[[metricCol]]
+        metricInfo[[metricCol]] <- NULL
+    }
+
+    ## Create annotations -----------------------------------------------------
+    rowAnnotRight <- ComplexHeatmap::rowAnnotation(
         Score = ComplexHeatmap::anno_barplot(
             rowAnnot[[scoreCol]],
             width = grid::unit(2, "cm"), 
             baseline = 0,
             axis = TRUE,
-            axis_param = list(side = "top"))
+            axis_param = list(side = "top")),
+        annotation_name_side = "top", 
+        annotation_name_rot = 0
     )
-    if (groupCol %in% colnames(colAnnot)) {
-        colAnnot <- ComplexHeatmap::columnAnnotation(
-            Weight = ComplexHeatmap::anno_barplot(
-                colAnnot[[weightCol]],
-                height = grid::unit(2, "cm"), 
-                baseline = 0),
-            Group = colAnnot[[groupCol]]
+    
+    if (!is.null(idInfo)) {
+        rowAnnotLeft <- ComplexHeatmap::rowAnnotation(
+            df = idInfo,
+            col = idColors
         )
     } else {
-        colAnnot <- ComplexHeatmap::columnAnnotation(
-            Weight = ComplexHeatmap::anno_barplot(
-                colAnnot[[weightCol]],
-                height = grid::unit(2, "cm"), 
-                baseline = 0)
-        )
+        rowAnnotLeft <- NULL
     }
     
-    heatmapCols = circlize::colorRamp2(
-        seq(min(mat), max(mat), length = 3), 
-        c("blue", "#EEEEEE", "red")
+    colAnnotTop <- ComplexHeatmap::columnAnnotation(
+        Weight = ComplexHeatmap::anno_barplot(
+            colAnnot[[weightCol]],
+            height = grid::unit(2, "cm"), 
+            baseline = 0),
+        annotation_name_side = "left",
+        annotation_name_rot = 90
     )
-    
+    if (!is.null(metricInfo)) {
+        colAnnotBottom <- ComplexHeatmap::columnAnnotation(
+            df = metricInfo,
+            col = metricColors
+        )
+    } else {
+        colAnnotBottom <- NULL
+    }
+
+    minmat <- min(mat, na.rm = TRUE)
+    maxmat <- max(mat, na.rm = TRUE)
+    if (minmat < 0) {
+        if (maxmat <= 0) {
+            heatmapCols = circlize::colorRamp2(
+                c(minmat, maxmat), 
+                c("blue", "#EEEEEE")
+            )
+        } else if (maxmat > 0) {
+            heatmapCols = circlize::colorRamp2(
+                c(minmat, 0, maxmat), 
+                c("blue", "#EEEEEE", "red")
+            )
+        }
+    } else {
+        heatmapCols = circlize::colorRamp2(
+            c(minmat, maxmat), 
+            c("#EEEEEE", "red")
+        )
+    }
+
     ComplexHeatmap::Heatmap(
         matrix = mat, name = "Relative\nvalue",
         col = heatmapCols,
-        na_col = "lightgrey",
+        na_col = "white",
```

Comment:
Is there a color which we (our minds AND the conventions) do connect somehow to being missing?
White is fine to me - as long as we do not "allow" it in the color palette?

Reply:
Yeah, not sure. Right now there's no white for the non-missing values (it's red, blue, #EEEEEE), so I thought white could be used for missing ones. Black felt too invasive.

Reply:
If we go 
- red-through-lightyellow-to-blue for the divergent palette
- a hue scale with increasing color (starting from a non-white level) for continuous, one-signed
- discrete ones for categories
I think white is a safe & sound choice for NA.

Other thought, how often do we expect that actually? Probably not so often... Or maybe I just don't have the right example in mind

Reply:
I agree - hopefully there shouldn't be too many NAs (also since the aggregated scores will be a bit difficult to interpret). I can think of situations where some methods don't return nominal p-values for example, or can't account for covariates.